src/Analysis/Ast/Impl/Analyzer/ModuleWalker.cs

```diff
@@ -242,57 +244,152 @@ public ModuleWalker(IServiceContainer services, IPythonModule module, PythonAst
-                        sourceType.TransferDocumentationAndLocation(stubType);
-                        // TODO: choose best type between the scrape and the stub. Stub probably should always win.
-                        var source = Eval.CurrentScope.Variables[v.Name]?.Source ?? VariableSource.Declaration;
-                        Eval.DeclareVariable(v.Name, v.Value, source);
-                    }
+                        break;
                 }
             }
         }
 
         private static bool IsStubBetterType(IPythonType sourceType, IPythonType stubType) {
-            // If stub says 'Any' but we have better type, keep the current type.
             if (stubType.IsUnknown()) {
+                // Do not use worse types than what is in the module.
+                return false;
+            }
+            if (sourceType.IsUnknown()) {
+                return true; // Anything is better than unknowns.
+            }
+            if (sourceType.DeclaringModule.ModuleType == ModuleType.Specialized) {
+                // Types in specialized modules should never be touched.
+                return false;
+            }
+            if (stubType.DeclaringModule is TypingModule && stubType.Name == "Any") {
+                // If stub says 'Any' but we have better type, keep the current type.
                 return false;
             }
-            return sourceType.IsUnknown() || !(stubType.DeclaringModule is TypingModule) || stubType.Name != "Any";
+            // Types should be compatible except it is allowed to replace function by a class.
+            // Consider stub of threading that replaces def RLock(): by class RLock().
+            // Similarly, in _json, make_scanner function is replaced by a class.
+            if (sourceType.MemberType == PythonMemberType.Function && stubType.MemberType == PythonMemberType.Class) {
+                return true;
+            }
+            return sourceType.MemberType == stubType.MemberType;
+        }
+
+        private static void TransferDocumentationAndLocation(IPythonType s, IPythonType d) {
+            if (s.IsUnknown() || d.IsBuiltin || s.IsBuiltin) {
+                return; // Do not transfer location of unknowns or builtins
+            }
+            // Documentation and location are always get transferred from module type
+            // to the stub type and never the other way around. This makes sure that
+            // we show documentation from the original module and goto definition
+            // navigates to the module source and not to the stub.
+            if (s != d && s is PythonType src && d is PythonType dst) {
+                // If type is a class, then doc can either come from class definition node of from __init__.
+                // If class has doc from the class definition, don't stomp on it.
+                var transferDoc = true;
+                if (src is PythonClassType srcClass && dst is PythonClassType dstClass) {
```

Comment:
We need to be able to tell if documentation came from class definition, from `__init__` or from the class base since we don't want lower priority sources stomping on the higher ones (i.e. class definition wins over `__init__`)